ethereum/spec/src/main/java/tech/pegasys/teku/spec/datastructures/eth1/Eth1Address.java

```diff
@@ -27,17 +32,93 @@ public class Eth1Address extends Bytes20 {
           .format("byte")
           .build();
 
-  public static final Eth1Address ZERO = new Eth1Address(Bytes.wrap(new byte[SIZE]));
+  public static final Eth1Address ZERO =
+      new Eth1Address("0x0000000000000000000000000000000000000000");
+
+  private final String address;
+
+  private Eth1Address(String value) {
+    super(Bytes.fromHexString(value));
+    this.address = value;
+  }
 
-  public Eth1Address(final Bytes bytes) {
-    super(bytes);
+  private static void checkAddress(String value) {
+    checkArgument(value.startsWith("0x"), "Eth1Address (%s) should start with \"0x\".", value);
+    Bytes bytes = Bytes.fromHexString(value);
+    checkArgument(
+        bytes.size() == SIZE, "Eth1Address should be 20 bytes, but was %s bytes.", bytes.size());
   }
 
   public static Eth1Address fromHexString(String value) {
-    return new Eth1Address(Bytes.fromHexString(value));
+    checkAddress(value);
+    Eth1Address address = new Eth1Address(value);
+    Eth1Address encodedAddress = fromHexStringWithChecksum(value);
+    checkArgument(
+        address.equals(encodedAddress),
+        "Eth1Address fails checksum:\n got: %s\n exp: %s",
+        address,
+        encodedAddress);
+    return address;
+  }
+
+  /**
+   * Produce an encoded address according to <a
+   * href="https://eips.ethereum.org/EIPS/eip-55">EIP-55</a>.
+   *
+   * @param value The string representation of an Ethereum address.
+   * @return The encoded address with mixed-case checksum.
+   */
+  public static Eth1Address fromHexStringWithChecksum(String value) {
+    checkAddress(value);
+    String lowerCaseAddrWithoutPrefix = value.substring(2).toLowerCase();
+    String hashString =
+        Hash.keccak256(Bytes.wrap(lowerCaseAddrWithoutPrefix.getBytes(StandardCharsets.US_ASCII)))
+            .toString();
+    String ret = "0x";
+    for (int i = 2; i < value.length(); i++) {
+      String letter = String.valueOf(hashString.charAt(i));
+      if (Integer.parseInt(letter, 16) >= 8) {
+        ret += Character.toUpperCase(value.charAt(i));
+      } else {
+        ret += Character.toLowerCase(value.charAt(i));
+      }
+    }
+    return new Eth1Address(ret);
   }
 
   public static DeserializableTypeDefinition<Eth1Address> getJsonTypeDefinition() {
     return ETH1ADDRESS_TYPE;
   }
+
+  @Override
+  public String toHexString() {
+    return address;
+  }
+
+  @Override
+  public boolean equals(final Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    final Eth1Address a = (Eth1Address) o;
+    return address.equals(a.address);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(address);
+  }
```

Comment:
We probably don't want to override equals and hashCode here so that we maintain the equality rules of `Bytes` which doesn't care about the particular implementation class.